File: native-modules/src/audio/android/AudioModule.java
package com.mr_bat.audio;

import android.media.MediaPlayer;
import android.util.SparseArray;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.io.IOException;

public class AudioModule extends ReactContextBaseJavaModule {
    private final ReactApplicationContext reactContext;
    private final SparseArray<MediaPlayer> players;
    private int nextId = 1;

    public AudioModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.players = new SparseArray<>();
    }

    @Override
    public String getName() {
        return "AudioModule";
    }

    @ReactMethod
    public void initialize(Promise promise) {
        try {
            // Clear any existing players
            for (int i = 0; i < players.size(); i++) {
                MediaPlayer player = players.valueAt(i);
                if (player != null) {
                    player.release();
                }
            }
            players.clear();
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("INIT_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void loadSound(String fileName, Promise promise) {
        try {
            String assetPath = "audio/" + fileName;
            MediaPlayer player = new MediaPlayer();
            player.setDataSource(reactContext.getAssets().openFd(assetPath));
            player.prepare();
            
            int id = nextId++;
            players.put(id, player);
            promise.resolve(id);
        } catch (IOException e) {
            promise.reject("LOAD_ERROR", "Failed to load sound: " + e.getMessage());
        }
    }

    @ReactMethod
    public void playSound(int soundId, Promise promise) {
        MediaPlayer player = players.get(soundId);
        if (player != null) {
            try {
                player.start();
                promise.resolve(null);
            } catch (IllegalStateException e) {
                promise.reject("PLAY_ERROR", e.getMessage());
            }
        } else {
            promise.reject("INVALID_ID", "Sound ID not found");
        }
    }

    @ReactMethod
    public void pauseSound(int soundId, Promise promise) {
        MediaPlayer player = players.get(soundId);
        if (player != null) {
            try {
                player.pause();
                promise.resolve(null);
            } catch (IllegalStateException e) {
                promise.reject("PAUSE_ERROR", e.getMessage());
            }
        } else {
            promise.reject("INVALID_ID", "Sound ID not found");
        }
    }

    @ReactMethod
    public void stopSound(int soundId, Promise promise) {
        MediaPlayer player = players.get(soundId);
        if (player != null) {
            try {
                player.stop();
                player.prepare();
                promise.resolve(null);
            } catch (Exception e) {
                promise.reject("STOP_ERROR", e.getMessage());
            }
        } else {
            promise.reject("INVALID_ID", "Sound ID not found");
        }
    }

    @ReactMethod
    public void releaseSound(int soundId, Promise promise) {
        MediaPlayer player = players.get(soundId);
        if (player != null) {
            player.release();
            players.remove(soundId);
            promise.resolve(null);
        } else {
            promise.reject("INVALID_ID", "Sound ID not found");
        }
    }

    @ReactMethod
    public void setVolume(int soundId, float volume, Promise promise) {
        MediaPlayer player = players.get(soundId);
        if (player != null) {
            try {
                player.setVolume(volume, volume);
                promise.resolve(null);
            } catch (IllegalStateException e) {
                promise.reject("VOLUME_ERROR", e.getMessage());
            }
        } else {
            promise.reject("INVALID_ID", "Sound ID not found");
        }
    }

    @ReactMethod
    public void getCurrentTime(int soundId, Promise promise) {
        MediaPlayer player = players.get(soundId);
        if (player != null) {
            try {
                int currentPosition = player.getCurrentPosition();
                promise.resolve(currentPosition / 1000.0);
            } catch (IllegalStateException e) {
                promise.reject("TIME_ERROR", e.getMessage());
            }
        } else {
            promise.reject("INVALID_ID", "Sound ID not found");
        }
    }
}

File: native-modules/src/audio/android/AudioModule.kt
package com.mr_bat.audio

import android.media.MediaPlayer
import com.facebook.react.bridge.*
import java.io.IOException

class AudioModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private val audioPlayers = mutableMapOf<Int, MediaPlayer>()
    private var nextSoundId = 1

    override fun getName() = "AudioModule"

    @ReactMethod
    fun initialize(promise: Promise) {
        try {
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("INIT_ERROR", e)
        }
    }

    @ReactMethod
    fun loadSound(fileName: String, promise: Promise) {
        try {
            val resourceId = reactApplicationContext.resources.getIdentifier(
                fileName.substringBeforeLast("."),
                "raw",
                reactApplicationContext.packageName
            )

            if (resourceId == 0) {
                throw IOException("Resource not found: $fileName")
            }

            val player = MediaPlayer.create(reactApplicationContext, resourceId)
            val soundId = nextSoundId++
            audioPlayers[soundId] = player
            promise.resolve(soundId)
        } catch (e: Exception) {
            promise.reject("LOAD_ERROR", e)
        }
    }

    @ReactMethod
    fun playSound(soundId: Int, promise: Promise) {
        try {
            val player = getPlayer(soundId)
            player.start()
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("PLAY_ERROR", e)
        }
    }

    @ReactMethod
    fun pauseSound(soundId: Int, promise: Promise) {
        try {
            val player = getPlayer(soundId)
            player.pause()
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("PAUSE_ERROR", e)
        }
    }

    @ReactMethod
    fun stopSound(soundId: Int, promise: Promise) {
        try {
            val player = getPlayer(soundId)
            player.stop()
            player.prepare()
            player.seekTo(0)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("STOP_ERROR", e)
        }
    }

    @ReactMethod
    fun releaseSound(soundId: Int, promise: Promise) {
        try {
            val player = audioPlayers.remove(soundId)
            player?.release()
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("RELEASE_ERROR", e)
        }
    }

    @ReactMethod
    fun setVolume(soundId: Int, volume: Float, promise: Promise) {
        try {
            val player = getPlayer(soundId)
            player.setVolume(volume, volume)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("VOLUME_ERROR", e)
        }
    }

    @ReactMethod
    fun getCurrentTime(soundId: Int, promise: Promise) {
        try {
            val player = getPlayer(soundId)
            promise.resolve(player.currentPosition / 1000.0)
        } catch (e: Exception) {
            promise.reject("TIME_ERROR", e)
        }
    }

    private fun getPlayer(soundId: Int): MediaPlayer {
        return audioPlayers[soundId] ?: throw IllegalArgumentException("Invalid sound ID: $soundId")
    }

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        audioPlayers.values.forEach { it.release() }
        audioPlayers.clear()
    }
}